log4j-core/src/main/java/org/apache/logging/log4j/core/config/ConfigurationSource.java

```diff
@@ -367,7 +357,7 @@ private static ConfigurationSource getConfigurationSource(URL url) {
             try {
                 if (file != null) {
                     return new ConfigurationSource(urlConnection.getInputStream(), FileUtils.fileFromUri(url.toURI()));
-                } else if (JAR.equals(url.getProtocol())) {
+                } else if (urlConnection instanceof JarURLConnection) {
```

Comment:
1. The connection implementation is chosen by ClassLoader. It may be different from java.net.JarURLConnection. See the exception provided in the PR description.
2. I see no reason to duplicate this check. Even if JarURLConnection will handle other protocol, no harm will be made by the workaround

Reply:
The problem I have is that this change means only JarURLConnection is supported even if something else is implementing the JAR protocol. Instead, we should be investigating why the ClassCastException is occurring and address that.

Reply:
The problem with `sun.net.www.protocol.jar.JarURLConnection` was that it leaked file descriptors upon each call to `getLastModified()` (cf. [LOG4J2-3423](https://issues.apache.org/jira/browse/LOG4J2-3423)).

I am not sure why IntelliJ decided not to extend the abstract class `java.net.JarURLConnection`, but I guess we don't need any workaround with classes that do not extend `java.net.JarURLConnection`.

BTW: this problem was already reported as [LOG4J2-3047](https://issues.apache.org/jira/browse/LOG4J2-3047).

Reply:
@rgoers: my point is that `urlConnection instanceof JarURLConnection` is probably the right thing to do here.

As for the JDK bug: [JDK-6956385](https://bugs.openjdk.java.net/browse/JDK-6956385) has been reported so many years ago, that it should probably be considered a feature by now.

Reply:
https://github.com/openjdk/jdk/pull/12871 should fix this.